File: get_alt_ids.m
function altIDs = get_alt_ids(t_studyID, t_subjID)
%% Config: project name translation table
pnTable = {{'innerspeech', 'InSp'}};

SL_DEMO_MAT_FN = '/speechlab/2/jtour/SID/SL_demos.mat';
USE_MAT_STUDY_CODE_PROJS = {'SDAP'};
STUDY_CODE_COL = 2;
DATA_DIR_NAME_COL = 3;

%%
altIDs = cell(1, 2);
altIDs{1} = [t_studyID, t_subjID];
altIDs{2} = [t_studyID, '_', t_subjID];

if length(t_subjID) >= 2 && isequal(t_subjID(1), 'S') && ~isnan(str2double(t_subjID(2)))
    bs = 1;
    altIDs{end + 1} = [t_studyID, t_subjID(2 : end)];
    altIDs{end + 1} = [t_studyID, '_', t_subjID(2 : end)];
else
    bs = 0;
end

if length(t_subjID) > 6 && isequal(t_subjID(1 : 3), 'SEQ') && ...
        (isequal(t_subjID(6), 'P') || isequal(t_subjID(6), 'C'))
    bSEQPDS = 1;
    altIDs{end + 1} = strrep(t_subjID, 'SEQ', 'SEQPDS');            
else
    bSEQPDS = 0;
end

if length(t_subjID) > 4 && isequal(t_subjID(1 : 3), 'FRS') && ...
        ~isnan(str2double(t_subjID(4 : end)))
    bFRS = 1;
    altIDs{end + 1} = t_subjID;
else
    bFRS = 0;
end

if length(t_subjID) > 5 && isequal(t_subjID(1 : 4), 'CCRS') && ...
        ~isnan(str2double(t_subjID(5 : end)))
    bCCRS = 1;
    altIDs{end + 1} = t_subjID;
else
    bCCRS = 0;
end

if length(t_subjID) > 4 && isequal(t_subjID(1 : 3), 'SEQ') && ...
        ~isnan(str2double(t_subjID(4 : end)))
    bSEQ = 1;
    altIDs{end + 1} = t_subjID;
else
    bSEQ = 0;
end

if ~isempty(fsic(USE_MAT_STUDY_CODE_PROJS, t_studyID))
    load(SL_DEMO_MAT_FN);
    assert(exist('N', 'var') == 1);
    assert(exist('T', 'var') == 1);
    
    bFound = 0;
    for i1 = 1 : size(T, 1)
        if ~isempty(fsic(altIDs, T{i1, DATA_DIR_NAME_COL}))
            bFound = 1;
            break;
        end
    end
    
    if bFound
        altIDs{end + 1} = T{i1, STUDY_CODE_COL};
        altIDs{end + 1} = [t_studyID, '_', T{i1, STUDY_CODE_COL}];
    end
end

%--- Find alternative project names from the pnTable ---% 
altStudyIDs = get_alt_study_ids(pnTable, t_studyID);
if ~isempty(altStudyIDs)
    for i2 = 1 : numel(altStudyIDs)
        altIDs{end + 1} = [altStudyIDs{i2}, t_subjID];
        altIDs{end + 1} = [altStudyIDs{i2}, '_', t_subjID];
        if bs
            altIDs{end + 1} = [altStudyIDs{i2}, t_subjID(2 : end)];
            altIDs{end + 1} = [altStudyIDs{i2}, '_', t_subjID(2 : end)];
        end
    end
end
        
return

%% Subroutines
function altStudyIDs = get_alt_study_ids(pnTable, t_studyID)
altStudyIDs = {};
for i1 = 1 : length(pnTable)
    if isequal(pnTable{i1}{1}, t_studyID)
        altStudyIDs = pnTable{i1}(2 : end);
    end
end

File: get_alt_ids.py

def get_alt_study_ids(pnTable, t_studyID):
    altStudyIDs = []

    for (i0, t_row) in enumerate(pnTable):
        if t_row[0] == t_studyID:
            for (j0, item) in enumerate(t_row[1:]):
                altStudyIDs.append(item)

    return altStudyIDs



def get_alt_ids(t_studyID, t_subjID):
    pnTable = [["innerspeech", "InSp"]]

    altIDs = []

    altIDs.append(t_studyID + t_subjID)
    altIDs.append(t_studyID + "_" + t_subjID)

    if t_subjID.startswith("S") and t_subjID[1:].isdigit():
        bs = True
        altIDs.append(t_studyID + t_subjID[1:])
        altIDs.append(t_studyID + "_" + t_subjID[1:])
    else:
        bs = False

    if len(t_subjID) > 6 and t_subjID.startswith("SEQ") and \
       (t_subjID[5] == "P" or t_subjID[5] == "C"):
        altIDs.append(t_subjID.replace("SEQ", "SEQPDS"))

    if len(t_subjID) > 4 and t_subjID.startswith("FRS") and \
       t_subjID[3:].isdigit():
        altIDs.append(t_subjID)
        
    if len(t_subjID) > 5 and t_subjID.startswith("CCRS") and \
       t_subjID[4:].isdigit():
        altIDs.append(t_subjID)

    if len(t_subjID) > 4 and t_subjID.startswith("SEQ") and \
       t_subjID[3:].isdigit():
        altIDs.append(t_subjID)

    #=== TODO: SDAP subjects ===#
    

    #=== Alternative study IDs ===#
    altStudyIDs = get_alt_study_ids(pnTable, t_studyID)
    if len(altStudyIDs) > 0:
        for (i2, t_altStudyID) in enumerate(altStudyIDs):
            altIDs.append(t_altStudyID + t_subjID)
            altIDs.append(t_altStudyID + "_" + t_subjID)
        
        if bs:
            altIDs.append(t_altStudyID + t_subjID[1:])
            altIDs.append(t_altStudyID + "_" + t_subjID[1:])
            
    return altIDs
